### ISSUE TITLE
Restrictions in Phantom Node Snapping

### ISSUE THREAD

**Issue by MoKob** [Project-OSRM/osrm-backend#2260]
The pure `valid | invalid` treatment in https://github.com/Project-OSRM/osrm-backend/blob/develop/include/engine/geospatial_query.hpp#L435-L460 can result in very strange and potentially unwanted behaviour.

On narrow bearing ranges, we will snap to the nearest location that is within the exact bearing range.
Imagine specifying a bearing of 180, turning out of your south facing driveway. Initial requests will search for the next south-facing road. This road can be far away, however.

I've tested this with the following result: Given a north bearing start of [this](http://map.project-osrm.org/?z=16&center=62.002783%2C-6.805944&loc=62.000149%2C-6.808155&loc=62.004541%2C-6.802533&hl=en&alt=0) route, the returned phantom node is the target of the route.

We should discuss ways of how we want to enable better phantom node discovery, given restriction parameters.

**Comment by TheMarex** [comment]
> Imagine specifying a bearing of 180, turning out of your south facing driveway. Initial requests will search for the next south-facing road. This road can be far away, however.

The real fix here is to not specify a narrow bearing range. If you are not sure the bearing is 100% aligned with what you want to snap to - reflect that uncertainty in the range parameter.

Implementing a "combined" metric of bearing and distance:

1. The StaticRTree query relies on the fact that the queue is sorted by something that corresponds to the distance.

2. There are two objects in the queue: `Rectangles` (inner tree nodes) and `Segments` (final leaves). You can't take the bearing of a Rectangle since they are always axis-aligned.

3. The result will not even be more intuitive: Every weighting formula that we could come up with is not easily verifiable by a human without cracking out a calculator. Whereas it is easy to explain that segments looked at from near to far and then filtered if the bearing is in `{bearing}+-{range}`.

4. If users want to avoid the "snap to segment 5km effect" they can simply specify a radius as well.

TL;DR; implementing this is not possible without a different query algorithm and the result will not necessarily be more intuitive.

**Comment by MoKob** [comment]
The problem I see here is that we find very unreasonable items based on the combination of restrictions.

 - We are only searching wide radiuses due to small components.
 - Restricting the bearing is only necessary to both find the correct direction and distinguish between otherwise valid candidates

The radius as a solution does not work fully, since we end up messing up the small-component part.

In my opinion, we should see that we only return a vertex far away, if the closer ones are small components.
